From 6368843396b2ffbda96903c4ec947956cfc6f1b5 Mon Sep 17 00:00:00 2001
From: xdBronch <51252236+xdBronch@users.noreply.github.com>
Date: Wed, 12 Nov 2025 10:12:07 -0500
Subject: [PATCH 059/133] fix 'redundant comptime keyword' error source
 location and add tests
X-Developer-Signature: v=1; a=openpgp-sha256; l=2690; i=xtex@astrafall.org;
 h=from:subject; bh=tJ01hsh7gjhMnYQf/i36Cy8ZKad/Urwm8GNh//AeMog=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78pg+Nt7a/8CNscVsffvJ+ieme21mqftNjZB6Nun3B
 +8D+q6BHRNZGMS4GCzFFFmKDBu8WXXS+UWXlcvCzGFlAhkiLdLAwMDAwMLAl5uYV2qkY6Rnqm2o
 Z2ioY6RjxMDFKQBTXeHN8M94ty6XRvPluEuCKSv37jgQ+r1J6PtUr1eZ7z9kGAR/uPSdkeHEGc+
 VMULdy+STk0u/8JmxXq1gnO7lGHu3kSNvwje/anYA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit 071453d5b9a8a48211c4517185df1b41d42107c7)
---
 lib/std/zig/AstGen.zig                                |  8 ++++----
 .../compile_errors/redundant_comptime_keyword.zig     | 11 +++++++++++
 2 files changed, 15 insertions(+), 4 deletions(-)
 create mode 100644 test/cases/compile_errors/redundant_comptime_keyword.zig

diff --git a/lib/std/zig/AstGen.zig b/lib/std/zig/AstGen.zig
index c208661359b4..da93086396b2 100644
--- a/lib/std/zig/AstGen.zig
+++ b/lib/std/zig/AstGen.zig
@@ -2108,10 +2108,10 @@ fn comptimeExprAst(
     node: Ast.Node.Index,
 ) InnerError!Zir.Inst.Ref {
     const astgen = gz.astgen;
+    const tree = astgen.tree;
     if (gz.is_comptime) {
-        try astgen.appendErrorNode(node, "redundant comptime keyword in already comptime scope", .{});
+        try astgen.appendErrorTok(tree.nodeMainToken(node), "redundant comptime keyword in already comptime scope", .{});
     }
-    const tree = astgen.tree;
     const body_node = tree.nodeData(node).node;
     return comptimeExpr2(gz, scope, ri, body_node, node, .comptime_keyword);
 }
@@ -3460,7 +3460,7 @@ fn assignDestructure(gz: *GenZir, scope: *Scope, node: Ast.Node.Index) InnerErro
 
     const full = tree.assignDestructure(node);
     if (full.comptime_token != null and gz.is_comptime) {
-        return astgen.appendErrorNode(node, "redundant comptime keyword in already comptime scope", .{});
+        return astgen.appendErrorTok(full.comptime_token.?, "redundant comptime keyword in already comptime scope", .{});
     }
 
     // If this expression is marked comptime, we must wrap the whole thing in a comptime block.
@@ -3516,7 +3516,7 @@ fn assignDestructureMaybeDecls(
 
     const full = tree.assignDestructure(node);
     if (full.comptime_token != null and gz.is_comptime) {
-        try astgen.appendErrorNode(node, "redundant comptime keyword in already comptime scope", .{});
+        try astgen.appendErrorTok(full.comptime_token.?, "redundant comptime keyword in already comptime scope", .{});
     }
 
     const is_comptime = full.comptime_token != null or gz.is_comptime;
diff --git a/test/cases/compile_errors/redundant_comptime_keyword.zig b/test/cases/compile_errors/redundant_comptime_keyword.zig
new file mode 100644
index 000000000000..8f1b1da9beee
--- /dev/null
+++ b/test/cases/compile_errors/redundant_comptime_keyword.zig
@@ -0,0 +1,11 @@
+comptime {
+    _ = comptime 0;
+}
+comptime {
+    comptime _, _ = .{ 0, 0 };
+}
+
+// error
+//
+// :2:9: error: redundant comptime keyword in already comptime scope
+// :5:5: error: redundant comptime keyword in already comptime scope
-- 
2.52.0

